modules/hashi-snapshots.nix

```
@@ -0,0 +1,416 @@
+        script = snapshotScripts.${hashiService} job;
+      in "${script}/bin/${name}";
+    };
+  };
+
+  snapshotScripts = let
+    mkSnapshotScript = {
+      hashiService,
+      job,
+      extraInputs,
+      snapshotCmd,
+      envPrep,
+      roleCmd,
+    }:
+      pkgs.writeShellApplication {
+        name = "${hashiService}-snapshot-${job}-script.sh";
+        runtimeInputs = with pkgs; [coreutils hostname nushell] ++ extraInputs;
+        text = ''
+          set -x
+
+          SNAP_NAME="$BACKUP_DIR/$HASHI_SERVICE-$(hostname)-$(date +"%Y-%m-%d_%H%M%SZ$BACKUP_SUFFIX").snap"
+
+          applyPerms () {
+            TARGET="$1"
+            PERMS="$2"
+
+            chown "$OWNER" "$TARGET"
+            chmod "$PERMS" "$TARGET"
+          }
+
+          takeSnapshot () {
+            if [ ! -d "$BACKUP_DIR" ]; then
+              mkdir -p "$BACKUP_DIR"
+              applyPerms "$BACKUP_DIR" "0700"
+            fi
+            ${snapshotCmd}
+            applyPerms "$SNAP_NAME" "0400"
+          }
+
+          ${envPrep}
+
+          if ${roleCmd}; then
+            ROLE="leader"
+          else
+            ROLE="replica"
+          fi
+
+          if [ "$ROLE" = "leader" ] && [ "$INCLUDE_LEADER" = "true" ]; then
+            takeSnapshot
+          elif [ "$ROLE" = "replica" ] && [ "$INCLUDE_REPLICA" = "true" ]; then
+            takeSnapshot
+          fi
+
+          # shellcheck disable=SC2016
+          nu -c '
+            ls $"($env.BACKUP_DIR)"
+            | where name =~ $"($env.BACKUP_SUFFIX).snap$"
+            | where type == file
+            | sort-by modified
+            | drop ${toString cfg.${hashiService}.${job}.backupCount}
+            | each {|f| rm $"($f.name)"; echo $"Deleted: ($f.name)"}
+          '
+        '';
+      };
+  in {
+    consul = job:
+      mkSnapshotScript {
+        inherit job;
+        hashiService = "consul";
+        extraInputs = with pkgs; [consul gnugrep];
+        snapshotCmd = ''consul snapshot save "$SNAP_NAME"'';
+        envPrep = "";
+        roleCmd = "consul info | grep -E '^\\s*leader\\s+=\\s+true$'";
+      };
+
+    nomad = job:
+      mkSnapshotScript {
+        inherit job;
+        hashiService = "nomad";
+        extraInputs = with pkgs; [jq nomad];
+        snapshotCmd = ''nomad operator snapshot save "$SNAP_NAME"'';
+        envPrep = ''
+          set +x
+          NOMAD_TOKEN="$(< ${hashiTokens.nomad-snapshot})"
+          export NOMAD_TOKEN
+          set -x
+
+          STATUS="$(nomad agent-info --json)"
+        '';
+        roleCmd = ''jq -e '(.stats.nomad.leader // "false") == "true"' <<< "$STATUS"'';
+      };
+
+    vault = job:
+      mkSnapshotScript {
+        inherit job;
+        hashiService = "vault";
+        extraInputs = with pkgs; [jq vault-bin];
+        snapshotCmd = ''vault operator raft snapshot save "$SNAP_NAME"'';
+        envPrep = ''
+          set +x
+          VAULT_TOKEN="$(< ${hashiTokens.vault})"
```

```suggestion
          VAULT_TOKEN=$(< ${hashiTokens.vault})
```

I think word split never happens here so quotes should not be needed. Same in line 275.